Plutarch/Pretty.hs

```diff
@@ -0,0 +1,218 @@
+{-# LANGUAGE PatternSynonyms #-}
+
+module Plutarch.Pretty (prettyTerm) where
+
+import Data.Bifunctor (first)
+import Data.List (foldl')
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.String (fromString)
+import Data.Text (Text)
+import qualified Data.Text as Txt
+
+import System.Random (
+  Random (randomR),
+  RandomGen,
+  mkStdGen,
+  uniformR,
+ )
+
+import Prettyprinter ((<+>))
+import qualified Prettyprinter as PP
+
+import Plutus.V1.Ledger.Scripts (Script (unScript))
+import qualified PlutusCore as PLC
+import UntypedPlutusCore (
+  DeBruijn (DeBruijn),
+  DefaultFun,
+  DefaultUni,
+  Index,
+  Program (_progTerm),
+  Term (Apply, Builtin, Constant, Delay, Error, Force, LamAbs, Var),
+ )
+import qualified UntypedPlutusCore as UPLC (Term)
+
+import Plutarch.Internal.Other (ClosedTerm, compile)
+
+prettyTerm :: ClosedTerm a -> PP.Doc ()
+prettyTerm x = prettyUPLC . _progTerm . unScript $ compile x
+
+{- This isn't suitable for pretty printing UPLC from any source. It's primarily suited for Plutarch output.
+
+Practically speaking though, it should work with any _idiomatic_ UPLC.
+-}
+prettyUPLC :: UPLC.Term DeBruijn DefaultUni DefaultFun () -> PP.Doc ()
+prettyUPLC = go False mempty $ mkStdGen 42
+  where
+    go :: RandomGen g => Bool -> Map Index Text -> g -> UPLC.Term DeBruijn DefaultUni DefaultFun () -> PP.Doc ()
+    go _ _ _ (Constant _ c) = PP.pretty c
+    go _ _ _ (Builtin _ b) = PP.pretty b
+    go _ _ _ (Error _) = "ERROR"
+    go _ nameMap _ (Var _ (DeBruijn x)) = case Map.lookup (x - 1) nameMap of
+      Just nm -> PP.pretty nm
+      Nothing -> error "impossible: free variable"
+    go _ nameMap g (Force _ t@Apply {}) = "!" <> PP.parens (go False nameMap g t)
+    go _ nameMap g (Force _ t@LamAbs {}) = "!" <> PP.parens (go False nameMap g t)
+    go _ nameMap g (Force _ t) = "!" <> go False nameMap g t
+    go _ nameMap g (Delay _ t@Apply {}) = "~" <> PP.parens (go False nameMap g t)
+    go _ nameMap g (Delay _ t@LamAbs {}) = "~" <> PP.parens (go False nameMap g t)
+    go _ nameMap g (Delay _ t) = "~" <> go False nameMap g t
+    go _ nameMap g (LamAbs _ _ t') =
+      PP.parens $
+        let (depth, t) = unwrapLamAbs 0 t'
+            (names, _, finalG) =
+              foldl'
+                ( \(l, existingNames, g) _ ->
+                    let (newName, newG) = freshVarName existingNames g
+                     in (newName : l, Set.insert newName existingNames, newG)
+                )
+                ([], Set.fromList $ Map.elems nameMap, g)
+                [0 .. depth]
+         in PP.hang indentWidth $
+              PP.sep
+                [ "\\" <> PP.hsep (reverse $ map PP.pretty names) <+> "->"
+                , go False (Map.mapKeys (+ (depth + 1)) nameMap <> Map.fromList (zip [0 .. depth] names)) finalG t
+                ]
+    go _ nameMap g (Apply _ (LamAbs _ _ t) arg) =
```

Comment:
I currently only do prettified let bindings for alternating `Apply` + `LamAbs` chains (i.e `Apply (LamAbs (Apply (LamAbs ...)))`). This is the kind of structure produced by `plet`.

Something like `Apply (Apply (Apply (LamAbs ...)))` will not produce let bindings, just a lambda and an application.

I think this conveys the intent better, and makes it easier to identify some optimizations (inlined hoisted funcs etc). But I understand that having let bindings for the other structure as well might make things easier to the eye.